src/libstd/time/duration.rs

```diff
@@ -168,15 +173,114 @@ impl Div<u32> for Duration {
 
 impl fmt::Display for Duration {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match (self.secs, self.nanos) {
-            (s, 0) => write!(f, "{}s", s),
-            (0, n) if n % NANOS_PER_MILLI == 0 => write!(f, "{}ms",
-                                                         n / NANOS_PER_MILLI),
-            (0, n) if n % 1_000 == 0 => write!(f, "{}µs", n / 1_000),
-            (0, n) => write!(f, "{}ns", n),
-            (s, n) => write!(f, "{}.{}s", s,
-                             format!("{:09}", n).trim_right_matches('0'))
+        // FIXME: fmt::Formatter::with_padding() isn't public, and the pad() function applies
+        // precision in a way we don't want. At such time as it becomes reasonable to support
+        // padding without reimplementing fmt::Formatter::with_padding() here, this should be
+        // updated to support padding. In preparation for that, we're already writing to a
+        // stack buffer.
+
+        // µ (U+00B5) is 0xC2B5 in UTF-8.
+        // The following is >= the longest possible string we can format.
+        let mut buf = *b"18446744073709551615.999999999\xC2\xB5s";
+
+        fn write_to_buf<'a>(dur: &Duration, f: &fmt::Formatter, mut buf: &'a mut [u8])
```

Comment:
This seems quite complicated, and perhaps not getting a lot of bang for our buck, would it be possible to use something simpler which perhaps has an intermediate buffer allocation? It's quite unlikely that the `fmt::Display` implementation for `Duration` is going to be at the middle of a tight loop, and otherwise this code is somewhat difficult to follow as there's a lot going on.